bin/i18n/translation_monitor/crowdin_validator.rb

```diff
@@ -0,0 +1,348 @@
+# A tool to download and validate translations made by specific users/vendors in Crowdin.
+# Can run frequently as a cronjob.
+#
+class CrowdinValidator
+  include CoreValidator
+  include DataIOHelper
+
+  CURRENT_DIR = File.dirname(__FILE__)
+  CONFIG_DIR = File.join(CURRENT_DIR, 'configs')
+  OUTPUT_DIR = File.join(CURRENT_DIR, 'outputs')
+
+  CROWDIN_CREDENTIAL_FILE = File.join(CONFIG_DIR, 'crowdin_v2_credential.json')
+  GOOGLE_DRIVE_CREDENTIAL_FILE = File.join(CONFIG_DIR, 'google_drive_credential.json')
+  CONFIG_FILE = File.join(CONFIG_DIR, 'crowdin_validator_config.json')
+  HISTORY_FILE = File.join(CONFIG_DIR, 'crowdin_validator_history.json')
+  LANGUAGE_FILE = File.join(CONFIG_DIR, 'crowdin_validator_languages.json')
+
+  REQUIRED_CONFIG_PARAMS = %w[
+    user_name
+    project_name
+    crowdin_language_id
+    start_date
+  ]
+
+  DATE_FORMAT = '%Y-%m-%d'
+
+  SourceString = Struct.new(:id, :project_id, :file_id, :identifier, :text, :is_hidden, :is_icu, :created_at, :updated_at, keyword_init: true)
+  Translation = Struct.new(:id, :string_id, :text, :crowdin_language_id, :user_id, :user_name, :created_at, keyword_init: true)
+
+  def initialize(credential_file = CROWDIN_CREDENTIAL_FILE)
+    api_token = read_from_json(credential_file)["api_token"]
+    @crowdin_client = Crowdin::Client.new do |config|
+      config.api_token = api_token
+    end
+    Dir.mkdir(OUTPUT_DIR) unless Dir.exist?(OUTPUT_DIR)
+  end
+
+  # Load and run all configurations from a config file.
+  # May combine with a history file to skip already processed data.
+  #
+  # This function has side effects. It may write to the local file system and Google Drive.
+  #
+  # @param config_file [String]
+  # @param history_file [String]
+  # @param update_history [Boolean]
+  # @param download_from_crowdin [Boolean]
+  #
+  def run_all_configs(config_file:, history_file: nil, update_history: false, download_from_crowdin: true)
+    configs = load_configs(config_file)
+
+    # Update configs using the history of last successful runs
+    history = {}
+    if history_file && File.exist?(history_file)
+      history = read_from_json(history_file)
+      merge_configs_and_history(configs, history)
+    end
+
+    # Execute all configs, one by one
+    configs.each do |config|
+      puts '=' * 50
+      puts "Executing config: #{config}"
+
+      begin
+        translations, source_strings, errors = run_config(config, download_from_crowdin)
+
+        # Update the history of last successful runs
+        if history_file && update_history
+          config_with_results = config.merge(
+            {
+              'results' => {
+                'translation_count' => translations.size,
+                'source_string_count' => source_strings.size,
+                'error_count' => errors.size
+              }
+            }
+          )
+          key = create_history_key(config)
+          history[key] = config_with_results
+          write_to_json(history, history_file)
```

Comment:
does the history_file have a bounded size? We don't want to have another file on the server that just keeps getting bigger until the FS is full.

Reply:
The size of the history file is bounded by the number of configurations. It saves only one record per configuration. I expect the number of configurations to be < 100.